rhel_osv/csaf.py

```diff
@@ -0,0 +1,144 @@
+# Copyright (c) 2024 Jason Shepherd
+# License: GPLv3+
+
+import json
+import logging
+from pathlib import Path
+from typing import Any
+
+logger = logging.Logger(__name__)
+
+
+class Reference(object):
+    """
+    class to handle CSAF References
+    """
+    def __init__(self, csaf_ref: dict[str, str]):
+        self.category = csaf_ref["category"]
+        self.summary = csaf_ref["summary"]
+        self.url = csaf_ref["url"]
+
+
+class Remediation(object):
+    """
+    class to handle remediation advise in CSAF data
+    """
+    def __init__(self, csaf_product_id: str, cpes: dict[str, str], purls: dict[str, str]):
+        if ":" not in csaf_product_id:
+            raise ValueError(f"Did not find ':' in product_id: {csaf_product_id}")
+        (self.product, self.product_version) = csaf_product_id.split(":", maxsplit=1)
+        self.cpe = cpes.get(self.product)
+        self.purl = purls.get(self.product_version)
+        # There are many pkg:oci/ remediations in Red Hat data. However there are no strict
+        # rules enforced on versioning Red Hat containers, therefore we cant compare container
+        # versions to each other with 100% accuracy at this time.
+        if not self.purl.startswith("pkg:rpm/"):
+            raise ValueError(f"Non RPM remediations are not supported in OSV at this time")
+
+        # NEVRA stands for Name Epoch Version Revision and Architecture
+        # We split the name from the rest of the 'version' data (EVRA). We store name as component.
+        split_component_version = self.product_version.rsplit("-", maxsplit=2)
+        if len(split_component_version) != 3:
+            raise ValueError(f"Could not convert component into NEVRA: {self.product_version}")
+        self.component = split_component_version[0]
+        self.fixed_version = "-".join((split_component_version[1], split_component_version[2]))
+
+
+class Vulnerability(object):
+    """
+    class to handle vulnerability information
+    """
+    def __init__(self, csaf_vuln: dict[str, Any], cpes: dict[str, str], purls: dict[str, str]):
+        self.cve_id = csaf_vuln["cve"]
+        for score in csaf_vuln["scores"]:
+            if "cvss_v3" in score:
+                self.cvss_v3_vector = score["cvss_v3"]["vectorString"]
+                self.cvss_v3_base_score = score["cvss_v3"]["baseScore"]
+        self.references = [Reference(r) for r in csaf_vuln["references"]]
+        self.remediations = []
+        for product_id in csaf_vuln["product_status"]["fixed"]:
+            try:
+                self.remediations.append(Remediation(product_id, cpes, purls))
+            except ValueError as e:
+                logger.warning(f"Could not parse product_id: {product_id}: {e}")
+
+
+
+def gen_dict_extract(key, var):
```

Comment:
I think we can get a [type hint](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#standard-duck-types) for `var` in here (requires importing `Iterable` from `types` as well)?

```suggestion
def gen_dict_extract(key, var: Iterable):
```

Reply:
Ok, added.